### DISCUSSION TITLE
Express + mikroorm + multer = entity manager updefined error

### DISCUSSION THREAD

**Discussion by NLin4com** [mikro-orm/mikro-orm#5759] [Q&A]
Here's the case: I have an app based on express. Microorm is configured and works properly in most of the cases, but when it comes to endpoints with file upload functionality something goes wrong. Locally everything works fine, but on staging I get error like `Cannot read properties of undefined (reading 'fork')` or something similar related to calling entity manager methods.

In `app.js` mikroorm is added like that:

```
const app = express()

(async () => {
    const em = (await orm).em;

    app.use((req, res, next) => {
        RequestContext.create(em, next);
    });

    app.listen(process.env.SERVER_PORT, () =>
        console.log(`Servers has been started on ${process.env.SERVER_PORT} port`)
    )
})()
```

Then here's full multer middleware code:

```
// Configure multer storage and file name
const storage = multer.diskStorage({
  destination: (req, file, cb) => {
    cb(null, "catalog/uploads/");
  },
  filename: (req, file, cb) => {
    const extIdx = file.originalname.lastIndexOf(".");
    const ext = file.originalname.substring(extIdx + 1);
    const name = file.originalname.substring(0, extIdx);
    cb(null, Date.now() + "-" + slugify(name) + "." + ext);
  }
});

// Create multer upload instance
const upload = multer({ storage: storage });

module.exports = {
  uploadMiddleware: upload.any()
};
```

And here's how it's applied to endpoint:

```
router.post('/upload-endpoint', [authMiddleware, uploadMiddleware], (req, res, next) => {
  uploadController.upload(req, res).catch((e) => errorMiddleware(e, req, res))
});
```

authMiddleware in this case contains code that goes to DB and retrieves a user by JWT. And this middleware ALWAYS works properly. But on the level of controller from time to time I get undefined error. Can anybody suggest how to deal with this error?

multer: 1.4.4-lts.1
mikro-orm: ^5.8.1

**Comment by NLin4com** [comment]
The only solution that works for now:
1. Move uploadMiddleware before authMiddleware (it's related to user entity and it's related entities that retrieved from DB and added to req object)
2. Recreate context in upload middleware callback:

```
module.exports = {
  uploadMiddleware: function (req, res, next) {
    const em = RequestContext.getEntityManager();
    upload.any()(req, res, () => {
      RequestContext.create(em, next);
    });
  }
};
```

But it doesn't look like really appropriate solution

**Reply by B4nan** [reply]
I guess the async context gets lost, there isn't much we can do. Changing the order of the middlewares is actually the correct solution.

The ORM middleware should be the last one before code that needs it, I don't know multer but the docs already mention the order being important when you have things like body parser and I can imagine this will be the same.

> We should register this middleware as the last one just before request handlers and before any of our custom middleware that is using the ORM. There might be issues when we register it before request processing middleware like queryParser or bodyParser, so definitely register the context after them.

https://mikro-orm.io/docs/identity-map#request-context

**Reply by NLin4com** [reply]
@B4nan maybe then you can suggest if it's possible to resolve an issue that I mentioned in p.1?

The thing is that I have an entity User with related entity Organization. It I don't switch order of middlewares related organization property would not be filled with data even after calling `wrap(this.organization).init()`. It looks like user entity looses connection with em and becomes unmanaged or something like that. Is there anything I can do with that?
